session/session.go

```diff
@@ -1921,16 +1908,25 @@ func (s *session) ExecutePreparedStmt(ctx context.Context, stmtID uint32, args [
 		return nil, errors.Errorf("invalid CachedPrepareStmt type")
 	}
 	executor.CountStmtNode(preparedStmt.PreparedAst.Stmt, s.sessionVars.InRestrictedSQL)
-	ok, err = s.IsCachedExecOk(ctx, preparedStmt)
+
+	var is infoschema.InfoSchema
+	if preparedStmt.ForUpdateRead {
```

Comment:
I think it's a bit risky currently to change the used schema version in transaction execution. Now the possible inconsistency issue bettwen the `current read` and `snapshot schema meta` is not resolved completely or it's worked around by a temprora index in https://github.com/pingcap/tidb/pull/22152/files. There are some implicit logic in tidb logic assuming the information schema is always consistent during the executions.

Reply:
No, it is not a change in this PR. It is change in #22381. I only extract the logic outside `IsCachedExecOK` and `optimize`, which is essentially a cleanup.

The real change is the removal of `tk3` and `is` in `ret := &amp;PreprocessorReturn{InfoSchema: is}`.

Reply:
The inforschema change for the cached plan short path is ok as it's only for autocommit point get read, the extraction may affect the common path in whch change inforschema version is risky.

Reply:
OK, it is not a must for me to change that anyway. It is reverted.